File: c/problem_24.c
#include "problem_24.h"
#include <stdio.h>
#include <stdlib.h>
#include <assert.h>

#define NULL_IDX (-1)
#define MAX_NODES 64

typedef struct {
    int val;
    int locked;
    int locked_desc_count;
    int parent;
    int left;
    int right;
} LockNode;

typedef struct {
    LockNode nodes[MAX_NODES];
    int count;
} LockTree;

static void ltInit(LockTree *t) { t->count = 0; }

static int ltAdd(LockTree *t, int val, int parent) {
    int idx = t->count++;
    t->nodes[idx].val              = val;
    t->nodes[idx].locked           = 0;
    t->nodes[idx].locked_desc_count = 0;
    t->nodes[idx].parent           = parent;
    t->nodes[idx].left             = NULL_IDX;
    t->nodes[idx].right            = NULL_IDX;
    return idx;
}

static int canLockOrUnlock(LockTree *t, int idx) {
    if (t->nodes[idx].locked_desc_count > 0) return 0;
    int p = t->nodes[idx].parent;
    while (p != NULL_IDX) {
        if (t->nodes[p].locked) return 0;
        p = t->nodes[p].parent;
    }
    return 1;
}

static int ltLock(LockTree *t, int idx) {
    if (t->nodes[idx].locked || !canLockOrUnlock(t, idx)) return 0;
    t->nodes[idx].locked = 1;
    int p = t->nodes[idx].parent;
    while (p != NULL_IDX) { t->nodes[p].locked_desc_count++; p = t->nodes[p].parent; }
    return 1;
}

static int ltUnlock(LockTree *t, int idx) {
    if (!t->nodes[idx].locked || !canLockOrUnlock(t, idx)) return 0;
    t->nodes[idx].locked = 0;
    int p = t->nodes[idx].parent;
    while (p != NULL_IDX) { t->nodes[p].locked_desc_count--; p = t->nodes[p].parent; }
    return 1;
}

void problem24(void) {
    printf("Problem 24:\n");
    LockTree t; ltInit(&t);
    int root  = ltAdd(&t, 1, NULL_IDX);
    int left  = ltAdd(&t, 2, root);
    int right = ltAdd(&t, 3, root);
    int ll    = ltAdd(&t, 4, left);
    t.nodes[root].left  = left;
    t.nodes[root].right = right;
    t.nodes[left].left  = ll;

    assert(ltLock(&t, left)   == 1);
    assert(t.nodes[left].locked == 1);
    assert(ltLock(&t, root)   == 0); /* locked descendant */
    assert(ltLock(&t, ll)     == 0); /* ancestor locked */
    assert(ltUnlock(&t, left) == 1);
    assert(t.nodes[left].locked == 0);
    assert(ltLock(&t, root)   == 1);
    printf("All Tests Passed\n");
}

File: typescript/problem_24.ts
class LockNode {
    val: number;
    locked: boolean = false;
    lockedDescCount: number = 0;
    parent: LockNode | null;
    left: LockNode | null = null;
    right: LockNode | null = null;

    constructor(val: number, parent: LockNode | null = null) {
        this.val = val;
        this.parent = parent;
    }

    isLocked(): boolean { return this.locked; }

    private canLockOrUnlock(): boolean {
        if (this.lockedDescCount > 0) return false;
        let p = this.parent;
        while (p) {
            if (p.locked) return false;
            p = p.parent;
        }
        return true;
    }

    lock(): boolean {
        if (this.locked || !this.canLockOrUnlock()) return false;
        this.locked = true;
        let p = this.parent;
        while (p) { p.lockedDescCount++; p = p.parent; }
        return true;
    }

    unlock(): boolean {
        if (!this.locked || !this.canLockOrUnlock()) return false;
        this.locked = false;
        let p = this.parent;
        while (p) { p.lockedDescCount--; p = p.parent; }
        return true;
    }
}

export function problem24(): void {
    console.log("Problem 24:");
    const root = new LockNode(1);
    root.left  = new LockNode(2, root);
    root.right = new LockNode(3, root);
    root.left.left = new LockNode(4, root.left);

    if (!root.left.lock())           throw new Error("assertion failed");
    if (!root.left.isLocked())       throw new Error("assertion failed");
    if (root.lock())                 throw new Error("assertion failed"); // locked descendant
    if (root.left.left.lock())       throw new Error("assertion failed"); // ancestor locked
    if (!root.left.unlock())         throw new Error("assertion failed");
    if (root.left.isLocked())        throw new Error("assertion failed");
    if (!root.lock())                throw new Error("assertion failed");
    console.log("All Tests Passed");
}
